### PR TITLE
improve type stabilities of `[try]parse`

### PR DESCRIPTION
**Pull request by aviatesk**

### CODE REVIEW

File: test/parse.jl [hunk 1]
```diff
@@ -298,6 +298,8 @@ end
     @test eltype([tryparse(Int, s, base=16) for s in String[]]) == Union{Nothing, Int}
     @test eltype([tryparse(Float64, s) for s in String[]]) == Union{Nothing, Float64}
     @test eltype([tryparse(Complex{Int}, s) for s in String[]]) == Union{Nothing, Complex{Int}}
+    # JET.@test_opt parse(UInt8, "12")
```
Comment (giordano) [inline test/parse.jl:301]: Leftovers?

Comment (aviatesk) [reply test/parse.jl:301]: Kind of. But I actually want to integrate JET into Julia Base soonish and enable it within a test suite. I'd like to turn on these leftovers then if people aren't unhappy with them.

File: base/parse.jl [hunk 2]
```diff
@@ -102,8 +102,9 @@ end
         _a <= _c <= _z ? _c-_a+a           : UInt32(base)
 end
 
-
-function tryparse_internal(::Type{T}, s::AbstractString, startpos::Int, endpos::Int, base_::Integer, raise::Bool) where T<:Integer
+# `@constprop :aggressive` here to make sure we const-prop' `raise` argument to obtain
+# the return-type stability for the `raise === true` case
+@constprop :aggressive function tryparse_internal(::Type{T}, s::AbstractString, startpos::Int, endpos::Int, base_::Integer, raise::Bool) where T<:Integer
```
Comment (Keno) [inline base/parse.jl:107]: If `raise` is that important for type stability, wouldn't it be better to lift it into the type domain? `@constprop :aggressive` on this function seems quite arbitrary.

Comment (aviatesk) [reply base/parse.jl]: Sounds reasonable. Lifted it to the type domain.

**Comment by KristofferC** [comment]
DelimitedFiles seems to use this internal function. We should fix that.

**Comment by aviatesk** [comment]
I would like to achieve this type information improvement by the compiler-side refinement (#46471).

**Pull request closed without merging**